src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp

```diff
@@ -63,7 +63,7 @@ bool ExecuteScalarSubqueriesMatcher::needChildVisit(ASTPtr & node, const ASTPtr
     if (node->as<ASTSelectQuery>())
     {
         /// Do not go to FROM, JOIN, UNION.
-        if (child->as<ASTTableExpression>() || child->as<ASTSelectQuery>())
+        if (child->as<ASTTableExpression>() || child->as<ASTSelectQuery>() || child->as<ASTTablesInSelectQuery>())
```

Comment:
Fix 1: ExecuteScalarSubqueriesVisitor was visiting the join elements of the query (contrary to the comment above) and altering the children of ASTTableExpression, but not the named pointers (using_expression_list, on_expression). This only worked because then we ignored children and kept using the named pointers.